base_layer/core/src/proof_of_work/randomx_factory.rs

```diff
@@ -0,0 +1,118 @@
+use crate::proof_of_work::monero_rx::MergeMineError;
+use randomx_rs::{RandomXCache, RandomXDataset, RandomXError, RandomXFlag, RandomXVM};
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex, RwLock},
+    time::Instant,
+};
+
+const MAX_VMS: usize = 5;
+
+#[derive(Default)]
+pub struct RandomXConfig {
+    pub use_large_pages: bool,
+}
+
+impl From<&RandomXConfig> for RandomXFlag {
+    fn from(source: &RandomXConfig) -> Self {
+        let mut result = RandomXFlag::default();
+        if source.use_large_pages {
+            result = result | RandomXFlag::FLAG_LARGE_PAGES
+        }
+        result
+    }
+}
+
+#[derive(Clone)]
+pub struct RandomXVMInstance {
+    // Note: If the cache and dataset drops, the vm will be wonky, so have to store all
+    // three for now
+    instance: Arc<Mutex<(RandomXVM, RandomXCache, RandomXDataset)>>,
```

Comment:
```suggestion
    instance: Arc<(RandomXVM, RandomXCache, RandomXDataset)>,
```

Reply:
I'm not 100% sure about that. The cache and the dataset don't change, but I'm not sure if the VM changes bits in FFI land. I would need to do a proper review of the `randomx_rs` crate. There are a few improvements that can be made there, e.g. pairing the lifetimes of the cache and dataset to the VM, which is not done right now. If the cache or dataset drop in rust, the memory will be released and the vm will continue happily.

I think the mutex is safer for now, and we can look at it again when I get some time to look at `randomx_rs`

Reply:
I suppose it is possible that unsafe could have been abused to allow unsynchronized mutation - I did test it before but let's stick with the safer approach.